lib/spack/spack/version.py

```diff
@@ -461,21 +463,28 @@ def __init__(self, string):
         if not isinstance(string, str):
             string = str(string)  # In case we got a VersionBase or GitVersion object
 
+        # An object that can lookup git refs to compare them to versions
+        self._ref_lookup = None
+        self.ref_version = None
+
         git_prefix = string.startswith('git.')
-        self.ref = string[4:] if git_prefix else string
+        pruned_string = string[4:] if git_prefix else string
+
+        if '=' in pruned_string:
+            self.ref, self.ref_version = pruned_string.split('=')
+        else:
+            self.ref = pruned_string
 
         self.is_commit = len(self.ref) == 40 and COMMIT_VERSION.match(self.ref)
         self.is_ref = git_prefix  # is_ref False only for comparing to VersionBase
         self.is_ref |= bool(self.is_commit)
 
         # ensure git.<hash> and <hash> are treated the same by dropping 'git.'
-        canonical_string = self.ref if self.is_commit else string
+        # unless we are assigning a version with =
+        canonical_string = self.ref if (self.is_commit and
```

Comment:
Is the goal here to be able to round-trip `@mytag=develop` without needing to use the `git` prefix? I think it might be preferable to require the git prefix in that context, but I'm open to discussion on it.

Reply:
Yes that is the goal.  I have tests for both. I wanted this to support any of the syntax that the hash-as-version supports.  My logic was that this feature is applicable in the exact same use cases as `package@hash` and `package@git.hash`

Reply:
If you have a strong preference to just support the `git` prefix then I'm fine with that.  I just think that since syntax and use case overlap is so similar to using the plain hash it may cause some user frustration/confusion in the future. But given that these use cases don't seem to be wildly popular yet it could be good to reign in the scope.  IDK there is a trade off here.